### PR TITLE
Introduce `-Zprecise-pre-release` unstable flag

### PR DESCRIPTION
**Pull request by eopb**
Tracking Issue: [#13290](https://github.com/rust-lang/cargo/issues/13290)

This change introduces the feature but does not yet attempt an implementation. The actual implementation will happen in future PRs.

r? @epage

### CODE REVIEW

File: tests/testsuite/precise_pre_release.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
+//! Tests for selecting pre-release versions with `update --precise`.
+
+use cargo_test_support::project;
+
+#[cargo_test]
+pub fn requires_nightly_cargo() {
```
Comment (weihanglo) [inline tests/testsuite/precise_pre_release.rs:6]: nit: `pub` is unnecessary.

```suggestion
fn requires_nightly_cargo() {
```

File: tests/testsuite/precise_pre_release.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+required by package `package v0.0.0 ([ROOT]/foo)`
+if you are looking for the prerelease package it needs to be specified explicitly
+    my-dependency = { version = "0.1.2-pre.0" }
+perhaps a crate was updated and forgotten to be re-vendored?"#,
+        )
+        .run()
+}
+
+#[cargo_test(nightly, reason = "-Zprecise_pre_release is unstable")]
+pub fn feature_exists() {
```
Comment (weihanglo) [inline tests/testsuite/precise_pre_release.rs:44]: ditto

```suggestion
fn feature_exists() {
```

File: tests/testsuite/precise_pre_release.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+location searched: `dummy-registry` index (which is replacing registry `crates-io`)
+required by package `package v0.0.0 ([ROOT]/foo)`
+if you are looking for the prerelease package it needs to be specified explicitly
+    my-dependency = { version = "0.1.2-pre.0" }
+perhaps a crate was updated and forgotten to be re-vendored?"#,
+        )
+        .run()
+}
+
+#[cargo_test(nightly, reason = "-Zprecise_pre_release is unstable")]
```
Comment (weihanglo) [inline tests/testsuite/precise_pre_release.rs:43]: This is for nightly rustc, but IIUC `-Zprecise_pre_release` is a cargo nightly feature and shouldn't require nightly rustc.

```suggestion
#[cargo_test]
```

File: tests/testsuite/precise_pre_release.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            [package]
+            name = "package"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("-Zprecise-pre-release update")
+        .masquerade_as_nightly_cargo(&["precise-pre-release"])
+        .with_status(0)
```
Comment (weihanglo) [inline tests/testsuite/precise_pre_release.rs:58]: 0 is the default.

```suggestion
```

**Review by weihanglo — COMMENTED** [review]
Thanks for the effort!

**Comment by eopb** [comment]
Thanks for your review @weihanglo, I've updated my commit with your suggestions

**Comment by epage** [comment]
@bors r+

Looks good to me, that @weihanglo's concerns have been resolved, and if any more crop up, they can be handled as this moves forward.

**Comment by ehuss** [comment]
Just FYI, it's a little confusing to me to have documentation for something that isn't implemented. Perhaps in the future there can be a warning indicating it isn't implemented yet?

**Pull request merged**